lib/internal/error-serdes.js

```diff
@@ -0,0 +1,111 @@
+'use strict';
+
+const Buffer = require('buffer').Buffer;
+const { serialize, deserialize } = require('v8');
+
+const kSerializedError = 0;
+const kSerializedObject = 1;
+const kInspectedError = 2;
+
+const GetPrototypeOf = Object.getPrototypeOf;
+const GetOwnPropertyDescriptor = Object.getOwnPropertyDescriptor;
+const GetOwnPropertyNames = Object.getOwnPropertyNames;
+const DefineProperty = Object.defineProperty;
+const Assign = Object.assign;
+const ObjectPrototypeToString =
+    Function.prototype.call.bind(Object.prototype.toString);
+const ForEach = Function.prototype.call.bind(Array.prototype.forEach);
+
+function TryGetAllProperties(object) {
+  const all = Object.create(null);
+  if (object === null)
+    return all;
+  Assign(all, TryGetAllProperties(GetPrototypeOf(object)));
+  const keys = GetOwnPropertyNames(object);
+  ForEach(keys, (key) => {
+    const descriptor = GetOwnPropertyDescriptor(object, key);
+    const getter = descriptor.get;
+    if (getter && key !== '__proto__') {
+      try {
+        descriptor.value = getter.call(object);
+      } catch (e) {}
+    }
+    if ('value' in descriptor && typeof descriptor.value !== 'function') {
+      delete descriptor.get;
+      delete descriptor.set;
+      all[key] = descriptor;
+    }
+  });
+  return all;
+}
+
+function GetConstructors(object) {
+  const constructors = [];
+
+  for (var current = object;
+    current !== null;
+    current = GetPrototypeOf(current)) {
+    const desc = GetOwnPropertyDescriptor(current, 'constructor');
+    if (desc && desc.value) {
+      DefineProperty(constructors, constructors.length, {
+        value: desc.value, enumerable: true
+      });
+    }
+  }
+
+  return constructors;
+}
+
+function GetName(object) {
+  const desc = GetOwnPropertyDescriptor(object, 'name');
+  return desc && desc.value;
+}
+
+let util;
+function lazyUtil() {
+  if (!util)
+    util = require('util');
+  return util;
+}
+
+function serializeError(error) {
+  try {
+    if (typeof error === 'object' &&
+        ObjectPrototypeToString(error) === '[object Error]') {
+      const constructors = GetConstructors(error);
+      for (var i = constructors.length - 1; i >= 0; i--) {
+        const name = GetName(constructors[i]);
+        if (name === 'Error' || name === 'TypeError' || name === 'RangeError') {
+          try { error.stack; } catch (e) {}
+          const serialized = serialize({
+            constructor: name,
+            properties: TryGetAllProperties(error)
+          });
+          return Buffer.concat([Buffer.from([kSerializedError]), serialized]);
+        }
+      }
+    }
+  } catch (e) {}
+  try {
+    const serialized = serialize(error);
+    return Buffer.concat([Buffer.from([kSerializedObject]), serialized]);
+  } catch (e) {}
+  return Buffer.concat([Buffer.from([kInspectedError]),
+                        Buffer.from(lazyUtil().inspect(error), 'utf8')]);
```

Comment:
What if inspect throws?

Reply:
In that case, we’re screwed and C++ reports that it can’t serialize the error:

```js
$ ./ayo -e 'new worker.Worker(`throw { inspect() { throw "foo"; } }`, { eval: true })'
internal/worker.js:115
    this.emit('error', new errors.Error('ERR_WORKER_UNSERIALIZABLE_ERROR'));
         ^

Error [ERR_WORKER_UNSERIALIZABLE_ERROR]: Serializing an uncaught exception failed
    at Worker.[kOnCouldNotSerializeErr] (internal/worker.js:115:24)
    at Worker.[kOnMessage] (internal/worker.js:136:45)
    at MessagePort.Worker.(anonymous function).on (internal/worker.js:95:23)
```

I think that’s reasonable?

Reply:
Yeah... that's fine I guess.